File: medium/0049_group_anagrams/solution.cpp
#include <array>
#include <map>
#include <string>
#include <vector>

class Solution {
public:
  std::vector<std::vector<std::string>>
  groupAnagrams(std::vector<std::string> &strs) {
    std::map<std::array<int, 26>, std::vector<std::string>> anagramsMap;

    for (const std::string &str : strs) {
      std::array<int, 26> charCounter{};
      for (const char &ch : str)
        ++charCounter[ch - 'a'];

      anagramsMap[charCounter].push_back(str);
    }

    std::vector<std::vector<std::string>> anagramGroups;
    anagramGroups.reserve(anagramsMap.size());

    for (const auto &[_, anagrams] : anagramsMap)
      anagramGroups.push_back(std::move(anagrams));

    return anagramGroups;
  }
};

File: medium/0049_group_anagrams/solution.py
from collections import defaultdict


class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        anagramsMap = defaultdict(list)

        for s in strs:
            anagramsMap[tuple(sorted(s))].append(s)

        return list(anagramsMap.values())
